api/pkg/apis/v1alpha1/managers/secrets/secrets-manager.go

```diff
@@ -0,0 +1,100 @@
+
+type SecretsManager struct {
+	managers.Manager
+	SecretProviders map[string]secret.ISecretProvider
+	Precedence      []string
+}
+
+func (s *SecretsManager) Init(context *contexts.VendorContext, cfg managers.ManagerConfig, providers map[string]providers.IProvider) error {
+	log.Debug(" M (secret): Init")
+	err := s.Manager.Init(context, cfg, providers)
+	if err != nil {
+		return err
+	}
+	s.SecretProviders = make(map[string]secret.ISecretProvider)
+	for key, provider := range providers {
+		if cProvider, ok := provider.(secret.ISecretProvider); ok {
+			s.SecretProviders[key] = cProvider
+		}
+	}
+	if val, ok := cfg.Properties["precedence"]; ok {
+		s.Precedence = strings.Split(val, ",")
+	}
+	if len(s.SecretProviders) == 0 {
+		log.Error(" M (secret): No secret providers found")
+		return v1alpha2.NewCOAError(nil, "No secret providers found", v1alpha2.BadRequest)
+	}
+	if len(s.SecretProviders) > 0 && len(s.Precedence) < len(s.SecretProviders) && len(s.SecretProviders) > 1 {
+		log.Error(" M (secret): Not enough precedence values")
+		return v1alpha2.NewCOAError(nil, "Not enough precedence values", v1alpha2.BadRequest)
+	}
+	for _, key := range s.Precedence {
+		if _, ok := s.SecretProviders[key]; !ok {
+			log.Error(" M (secret): Invalid precedence value: %s", key)
+			return v1alpha2.NewCOAError(nil, fmt.Sprintf("Invalid precedence value: %s", key), v1alpha2.BadRequest)
+		}
+	}
+	return nil
+}
+func (s *SecretsManager) Get(object string, field string, localContext interface{}) (string, error) {
+	log.Debugf(" M (secret): Get %v, secret provider size %d", object, len(s.SecretProviders))
+	if strings.Index(object, "::") > 0 {
+		parts := strings.Split(object, "::")
+		if len(parts) != 2 {
+			log.Errorf(" M (secret): Invalid object: %s", object)
+			return "", v1alpha2.NewCOAError(nil, fmt.Sprintf("Invalid object: %s", object), v1alpha2.BadRequest)
+		}
+		if provider, ok := s.SecretProviders[parts[0]]; ok {
+			if field == "" {
+				log.Errorf(" M (secret): field is empty")
+				return "", v1alpha2.NewCOAError(nil, "Field is empty", v1alpha2.BadRequest)
+			} else {
+				return provider.Read(parts[1], field, localContext)
+			}
+		}
+		log.Errorf(" M (secret): Invalid provider: %s", parts[0])
+		return "", v1alpha2.NewCOAError(nil, fmt.Sprintf("Invalid provider: %s", parts[0]), v1alpha2.BadRequest)
+	}
+	if len(s.SecretProviders) == 1 {
+		for _, provider := range s.SecretProviders {
+			if field == "" {
+				log.Errorf(" M (secret): field is empty")
+				return "", v1alpha2.NewCOAError(nil, "Field is empty", v1alpha2.BadRequest)
+			} else {
+				return provider.Read(object, field, localContext)
+			}
+		}
+	}
+	for _, key := range s.Precedence {
+		if provider, ok := s.SecretProviders[key]; ok {
+			if field == "" {
+				log.Errorf(" M (secret): field is empty")
+				return "", v1alpha2.NewCOAError(nil, "Field is empty", v1alpha2.BadRequest)
+			} else {
+				return provider.Read(object, field, localContext)
```

Comment:
I am not expecting precedence works in this way. If the first provider doesn't get the secret and field, shall we check it in the second provider and then the third provider?

If you only look up the secret in the first provider if provider is not specified, you can just define default provider rather than a list of Precedence. Right?